pkg/netns/netns_linux.go

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package netns
+
+import (
+	"fmt"
+	"os"
+	"runtime"
+
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sys/unix"
+)
+
+type NetNS struct {
+	fd int
+}
+
+// CreateNewNetns creates a network namespace and returns a handle to it.
+//
+// The current thread will not be moved to this new network namespace as with a basic
+// UNSHARE syscall - the original network namespace will be retained and a handle to the
+// new network namespace will be returned.
+//
+// The namespace created by this call is NOT pinned, or bind-mounted, and so will be
+// closed/lost when the process terminates.
+func New() (*NetNS, error) {
+	var innerFD int
+
+	g := new(errgroup.Group)
+	// Force this into a dedicated + threadlocked goroutine to avoid disturbing the caller's
+	// thread.
+	g.Go(func() error {
+		restore, err := saveRestoreCurrentThreadNetNS()
+		if err != nil {
+			return fmt.Errorf("Unable to obtain and lock current netns: %w", err)
+		}
+		defer restore()
+
+		// This will move our current (locked) thread to a new netNS.
+		// Once we get the FD of that new netNS, invoking the func
+		// returned by `saveRestoreCurrentThreadNetNS` will move us
+		// back to the original thread
+		err = unshare()
+		if err != nil {
+			return fmt.Errorf("Unable to create new netns: %w", err)
+		}
+		innerFD, err = getCurrent()
+		if err != nil {
+			return fmt.Errorf("Unable to get current netns: %w", err)
+		}
+		return nil
+	})
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	ns := &NetNS{fd: innerFD}
+	// This is not foolproof - it does not guarantee that the finalizer
+	// will be invoked, or when. Just that the go runtime will try to invoke it
+	// at some point after the `ns` instance goes out of scope, if the caller
+	// forgets to explicitly Close()
+	runtime.SetFinalizer(ns, func(f *NetNS) {
```

Comment:
I think this technically works, since it checks all the boxes in the SetFinalizer documentation, but this runs the risk of someone coming in later and closing over another variable in this function. We can pass a function pointer, like so:

```suggestion
	runtime.SetFinalizer(ns, (*NetNS).Close)
```

A method's receiver is implemented as the function's first argument under the hood.

Reply:
Ah, cheeky, ty.